File: trunk/Java/OPSJLib/src/ops/Publisher.java
/**
 *
 * Copyright (C) 2006-2009 Anton Gravestam.
 *
 * This file is part of OPS (Open Publish Subscribe).
 *
 * OPS (Open Publish Subscribe) is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * OPS (Open Publish Subscribe) is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with OPS (Open Publish Subscribe).  If not, see <http://www.gnu.org/licenses/>.
 */
package ops;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.logging.Level;
import java.util.logging.Logger;
import ops.archiver.OPSArchiverOut;
import ops.protocol.OPSMessage;
import ops.transport.inprocess.InProcessTransport;

/**
 *
 * @author Anton Gravestam
 */
public class Publisher
{

    private Topic topic;
    private int currentPublicationID = 1;
    private String name = "";
    private String key = "";
    private int reliableWriteNrOfResends = 1;
    private int reliableWriteTimeout = 1000;
    private byte[] bytes;
    private ByteBuffer buffer;
    private Participant participant;
    private SendDataHandler sendDataHandler;
    private InProcessTransport inProcessTransport;
    private long lastPublishTime;
    private volatile long sampleTime1;
    private volatile long sampleTime2;
    private boolean started = false;

    public Publisher(Topic topic)
    {
        this.topic = topic;
        bytes = new byte[topic.getSampleMaxSize()];
        buffer = ByteBuffer.allocateDirect(topic.getSampleMaxSize());

        this.participant = Participant.getInstance(topic.getDomainID(), topic.getParticipantID());
        inProcessTransport = participant.getInProcessTransport();
        init();
    }

    @Override
    protected void finalize() throws Throwable {
        super.finalize();
        // Must tell the sendDataHandler that we don't need it anymore
        // in case user hasn't called stop()
        stop();
        sendDataHandler = null;
    }

    private void init()
    {
        try
        {
            sendDataHandler = participant.getSendDataHandler(topic);
            start();
        }
        catch (CommException ex)
        {
            participant.report(this.getClass().getName(), "init", ex.getMessage());
        }
    }

    /// <summary>
    /// Start the publisher (necessary when using TCP as transport and it has been stopped)
    /// </summary>
    public synchronized void start()
    {
        if (!started)
        {
            // Tell the sendDataHandler that we need it
            sendDataHandler.addPublisher(this);
            started = true;
        }
    }

    /// <summary>
    /// Close down the publisher (necessary when using TCP as transport)
    /// </summary>
    public synchronized void stop()
    {
        if (started)
        {
            // Tell the sendDataHandler that we don't need it anymore
            sendDataHandler.removePublisher(this);
            started = false;
        }
    }

    protected void write(OPSObject o)
    {
        if (sendDataHandler == null)
        {
            init();
        }

        if (sendDataHandler == null)
        {
            return;
        }
        sampleTime2 = sampleTime1;
        sampleTime1 = System.currentTimeMillis();


        OPSMessage message = new OPSMessage();
        o.setKey(key);
        message.setData(o);
        message.setPublicationID(currentPublicationID);
        message.setTopicName(topic.getName());
        message.setPublisherName(name);

        //inProcessTransport.copyAndPutMessage(message);

        WriteByteBuffer buf = new WriteByteBuffer(buffer, StaticManager.MAX_SIZE);
        try
        {

            //buf.writeProtocol();
            //buf.write("");
            //buf.write(1);
            //buf.write(0);

            buffer.position(0);
            OPSArchiverOut archiverOut = new OPSArchiverOut(buf);

            archiverOut.inout("message", message);

            //If o has spare bytes, write them to the end of the buf
            if (o.spareBytes.length > 0)
            {
                buf.write(o.spareBytes, 0, o.spareBytes.length);
            }
            //Finish will in nrOf segments in all segments.
            buf.finish();
            int sizeToSend = buffer.position();

            //transport.send(archiverOut.getBytes());
            buffer.position(0);
            buffer.get(bytes);
            sendDataHandler.sendData(bytes, sizeToSend, topic);
            //sendDataHandler.sendData(buffer.array(), buf.position(), topic);

        }
        catch (IOException ex)
        {
            Logger.getLogger(Publisher.class.getName()).log(Level.SEVERE, null, ex);
            ex.printStackTrace();
        }
        incCurrentPublicationID();


    }

    

    public void writeAsOPSObject(OPSObject o)
    {
        write(o);
    }

    private void incCurrentPublicationID()
    {
        if (currentPublicationID < 10000)
        {
            currentPublicationID++;
        }
        else
        {
            currentPublicationID = 1;
        }
    }

    public String getName()
    {
        return name;
    }

    public void setName(String name)
    {
        this.name = name;
    }

    public Topic getTopic()
    {
        return topic;
    }

    public String getKey()
    {
        return key;
    }

    public void setKey(String key)
    {
        this.key = key;
    }

    public int getCurrentPublicationID()
    {
        return currentPublicationID;
    }

    public int getReliableWriteTimeout()
    {
        return reliableWriteTimeout;
    }

    public double getOutboundRate()
    {
        double sampleRate = 1.0/((sampleTime1 - sampleTime2) / 1000.0);
        double fakeRate  = 1.0/((System.currentTimeMillis() - sampleTime1) / 1000.0);

        if(sampleRate < fakeRate)
        {
            return sampleRate;
        }
        else
        {
            return fakeRate;
        }
    }

    
    

    public void setReliableWriteTimeout(int reliableWriteTimeout)
    {
        this.reliableWriteTimeout = reliableWriteTimeout;
    }

    public int getReliableWriteNrOfResends()
    {
        return reliableWriteNrOfResends;
    }

    public void setReliableWriteNrOfResends(int reliableWriteNrOfResends)
    {
        this.reliableWriteNrOfResends = reliableWriteNrOfResends;
    }

}

File: C#/Ops/Publisher.cs
///////////////////////////////////////////////////////////
//  Publisher.cs
//  Implementation of the Class Publisher
//  Created on:      12-nov-2011 09:25:34
//  Author:
///////////////////////////////////////////////////////////

using System;
using System.IO;

namespace Ops 
{
	public class Publisher 
    {
		private byte [] bytes;
		private long currentPublicationID = 0;
		private InProcessTransport inProcessTransport;
		private string key = "";
		private string name = "";
		private Participant participant;
		private int reliableWriteNrOfResends = 1;
		private int reliableWriteTimeout = 1000;
        private long sampleTime1;
        private long sampleTime2;
        private ISendDataHandler sendDataHandler;
		private Topic topic;
        private bool started = false;

        public Publisher(Topic topic)
        {
            this.topic = topic;
            this.bytes = new byte[topic.GetSampleMaxSize()];

            this.participant = Participant.GetInstance(topic.GetDomainID(), topic.GetParticipantID());
            this.inProcessTransport = participant.GetInProcessTransport();
            Init();
        }

        ~Publisher()
        {
            // Must tell the sendDataHandler that we don't need it anymore
            Stop();
            this.sendDataHandler = null;
            this.participant = null;
        }

        private void Init()
        {
            try
            {
                this.sendDataHandler = this.participant.GetSendDataHandler(topic);
                Start();
            }
            catch (Exception ex)
            {
                Logger.ExceptionLogger.LogMessage(this.GetType().Name + " Init: " + ex.ToString());
            }
        }

        /// <summary>
        /// Start the publisher (necessary when using TCP as transport and it has been stopped)
        /// </summary>
        public void Start()
        {
            if (!started)
            {
                // Tell the sendDataHandler that we need it
                this.sendDataHandler.AddPublisher(this);
                started = true;
            }
        }

        /// <summary>
        /// Close down the publisher (necessary when using TCP as transport)
        /// </summary>
        public void Stop()
        {
            if (started)
            {
                // Tell the sendDataHandler that we don't need it anymore
                this.sendDataHandler.RemovePublisher(this);
                started = false;
            }
        }

        /// <summary>
        /// Get IP and port used as endpoint when sending a message
        /// </summary>
        /// <param name="IP"></param>
        /// <param name="port"></param>
        public void GetLocalEndpoint(ref string IP, ref int port)
        {
            this.sendDataHandler.GetLocalEndpoint(ref IP, ref port);
        }

        public void CheckTypeString(string typeString)
        {
            if (topic.GetTypeID() != typeString)
            {
                throw new OPSInvalidTopicException("Topic type string mismatch!");
            }
        }

        protected void Write(OPSObject opsObject)
        {
            if (opsObject == null)
            {
                throw new CommException("Trying to send OPSObject that is null!");
            }

            if (sendDataHandler == null)
            {
                Init();
            }

            if (sendDataHandler == null)
            {
                return;
            }

            this.sampleTime2 = this.sampleTime1;
            this.sampleTime1 = System.DateTime.Now.Ticks;

            OPSMessage message = new OPSMessage();
            opsObject.SetKey(this.key);
            message.SetData(opsObject);
            message.SetPublicationID(this.currentPublicationID);
            message.SetTopicName(this.topic.GetName());
            message.SetPublisherName(this.name);

            //inProcessTransport.copyAndPutMessage(message);
            WriteByteBuffer buf = new WriteByteBuffer(this.bytes, Globals.MAX_SEGMENT_SIZE);
            try
            {
                OPSArchiverOut archiverOut = new OPSArchiverOut(buf);

                archiverOut.Inout("message", message);

                // If o has spare bytes, write them to the end of the buf
                if (opsObject.spareBytes.Length > 0)
                {
                    buf.Write(opsObject.spareBytes, 0, opsObject.spareBytes.Length);
                }
                // Finish will fill in nrOf segments in all segments.
                buf.Finish();
                int sizeToSend = buf.Position();

                sendDataHandler.SendData(this.bytes, sizeToSend, this.topic);
            }
            catch (System.IO.IOException ex)
            {
                Logger.ExceptionLogger.LogException(ex);
            }
            IncCurrentPublicationID();
        }

        public void WriteAsOPSObject(OPSObject opsObject)
        {
            Write(opsObject);
        }

        private void IncCurrentPublicationID()
        {
            if (this.currentPublicationID < long.MaxValue)
            {
                this.currentPublicationID++;
            }
            else
            {
                this.currentPublicationID = 0;
            }
        }

        public string GetName()
        {
            return this.name;
        }

        public void SetName(string name)
        {
            this.name = name;
        }

        public Topic GetTopic()
        {
            return this.topic;
        }

        public string GetKey()
        {
            return this.key;
        }

        public void SetKey(string key)
        {
            this.key = key;
        }

        public long GetCurrentPublicationID()
        {
            return this.currentPublicationID;
        }

        public int GetReliableWriteTimeout()
        {
            return this.reliableWriteTimeout;
        }

        public double GetOutboundRate()
        {
            double sampleRate = 1.0 / ((this.sampleTime1 - this.sampleTime2) / 10000000.0);
            double fakeRate = 1.0 / ((System.DateTime.Now.Ticks - this.sampleTime1) / 10000000.0);

            if (sampleRate < fakeRate)
            {
                return sampleRate;
            }
            else
            {
                return fakeRate;
            }
        }

        public void SetReliableWriteTimeout(int reliableWriteTimeout)
        {
            this.reliableWriteTimeout = reliableWriteTimeout;
        }

        public int GetReliableWriteNrOfResends()
        {
            return this.reliableWriteNrOfResends;
        }

        public void SetReliableWriteNrOfResends(int reliableWriteNrOfResends)
        {
            this.reliableWriteNrOfResends = reliableWriteNrOfResends;
        }

	}

}